beacon-chain/sync/rpc_blob_sidecars_by_range.go

```diff
@@ -0,0 +1,179 @@
+	"go.opencensus.io/trace"
+)
+
+func (s *Service) streamBlobBatch(ctx context.Context, batch blockBatch, stream libp2pcore.Stream) (uint64, error) {
+	ctx, span := trace.StartSpan(ctx, "sync.streamBlobBatch")
+	defer span.End()
+	var writes uint64
+	for _, b := range batch.canonical() {
+		root := b.Root()
+		scs, err := s.cfg.beaconDB.BlobSidecarsByRoot(ctx, b.Root())
+		if errors.Is(err, db.ErrNotFound) {
+			continue
+		}
+		if err != nil {
+			s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+			return writes, errors.Wrapf(err, "could not retrieve sidecars for block root %#x", root)
+		}
+		for _, sc := range scs {
+			SetStreamWriteDeadline(stream, defaultWriteDuration)
+			if chunkErr := WriteBlobSidecarChunk(stream, s.cfg.chain, s.cfg.p2p.Encoding(), sc); chunkErr != nil {
+				log.WithError(chunkErr).Debug("Could not send a chunked response")
+				s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+				tracing.AnnotateError(span, chunkErr)
+				return writes, chunkErr
+			}
+			s.rateLimiter.add(stream, 1)
+			writes += 1
+		}
+	}
+	return writes, nil
+}
+
+// blobsSidecarsByRangeRPCHandler looks up the request blobs from the database from a given start slot index
+func (s *Service) blobSidecarsByRangeRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
+	ctx, span := trace.StartSpan(ctx, "sync.BlobsSidecarsByRangeHandler")
+	defer span.End()
+	ctx, cancel := context.WithTimeout(ctx, respTimeout)
+	defer cancel()
+	SetRPCStreamDeadlines(stream)
+	log := log.WithField("handler", p2p.BlobSidecarsByRangeName[1:]) // slice the leading slash off the name var
+
+	r, ok := msg.(*pb.BlobSidecarsByRangeRequest)
+	if !ok {
+		return errors.New("message is not type *pb.BlobsSidecarsByRangeRequest")
+	}
+	if err := s.rateLimiter.validateRequest(stream, 1); err != nil {
+		return err
+	}
+	rp, err := validateBlobsByRange(r, s.cfg.chain.CurrentSlot())
+	if err != nil {
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(stream.Conn().RemotePeer())
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	// Ticker to stagger out large requests.
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	batcher, err := newBlockRangeBatcher(rp, s.cfg.beaconDB, s.rateLimiter, s.cfg.chain.IsCanonical, ticker)
+
+	var batch blockBatch
+	var totalWrites uint64
+	for batch, ok = batcher.next(ctx, stream); ok; batch, ok = batcher.next(ctx, stream) {
+		batchStart := time.Now()
+		writes, err := s.streamBlobBatch(ctx, batch, stream)
+		rpcBlobsByRangeResponseLatency.Observe(float64(time.Since(batchStart).Milliseconds()))
+		if err != nil {
+			return err
+		}
+		totalWrites += writes
+		// once we have written MAX_REQUEST_BLOB_SIDECARS, we're done serving the request
+		if totalWrites >= params.BeaconNetworkConfig().MaxRequestBlobsSidecars {
```

Comment:
I noticed this is checked after we have streamed the batch, would it be an issue if we streamed an excessive number of sidecars to the peer (beyond the limit).

Reply:
Good catch. I rewrote the bookkeeping code to stop streaming results as soon as we hit the limit: https://github.com/prysmaticlabs/prysm/pull/12420/commits/80555217f2a811f8ad00b1db586f592ab92707d7